ggml-common.h

```diff
@@ -137,6 +137,13 @@ typedef sycl::half2 ggml_half2;
 
 #endif // GGML_COMMON_DECL_CUDA || GGML_COMMON_DECL_HIP
 
+#define QK2_2 32
+typedef struct {
+    ggml_half d;           // delta
```

Comment:
Looks like a more resonable choice is remove the scale, but it will trigger an error in `VALIDATE_ROW_DATA_D_F16_IMPL(block_q2_2, data, nb)` since there is no scale in q2_2, should I just do no validating to the no-scale q2_2 block,?

Reply:
If there are no invalid values, no validating needs to be done.

But it seems not all of the 256 possible values of a byte are used for 4 packed ternary numbers in `Q2_2`, so maybe these invalid values could be detected by checking whether non-zero bytes map to zero in `q22_grid`.

Reply:
The values are not used by Bitnet, but a different model could use them (e.g. TwoBitnet `(-2, -1, 0, 1)`), so technically they are not invalid

Reply:
The non-scale Q2_2 block is commited, please review.

Reply:
I am not sure how feasible it would be to reuse the same format with a different mapping, since the dot product function will likely make assumptions about the mapping. Which means that using a different mapping will require a different dot product function, which will require a different ggml type id. I think it would be good to verify that the unassigned value is never used in the tensor validation.

Reply:
Yes, it's not feasible to have different mappings. I'm suggesting to reuse the same `Q2_K` mapping, because I don't think we gain anything from the new ternary mapping, unless it starts packing the data in less than 2 bits. This would require whole new kernels to be implemented, while with the existing mapping, the code should be mostly copy-paste + potentially deduplicated in the future and has the benefit to support actual 2-bit models/quantizations

I could be missing something so definitely let me know if there are concerns or doubts

Reply:
So if I understand correctly, then it would be a 2-bit type without block scales, that would be defined to have a fixed offset of -2?

Reply:
> since the dot product function will likely make assumptions about the mapping

> that would be defined to have a fixed offset of -2?

Currently, it seems like each 2-bit value in Q2_2 can be easily decoded by sign-extending it.

| 2 Bits | Byte | Number |
|--------|------|--------|
| `0b00` |`0x00`|      0 |
| `0b01` |`0x01`|      1 |
| `0b10` |`0xFE`|     -2 |
| `0b11` |`0xFF`|     -1 |

Currently, the unused value in Q2_2 is `0b10`, so some `0x00000000` values in `q22_grid` would need to be set correctly if it's used for `-2` and to keep the equivalence with sign-extension.

> I don't think we gain anything from the new ternary mapping, unless it starts packing the data in less than 2 bits. This would require whole new kernels to be implemented

Actually, I have some prototype of a `1.625 bpw` ternary packing scheme (and AVX2 `vec_dot`) which I'll likely push somewhere. Currently it's only very slightly slower (<1 GB/s float32 throughput of difference) than the AVX2 Q2_2 when both use a lookup table.

Reply:
Yes, same way as how `Q4_0` has a fixed offset of -8

Reply:
> I have some prototype of a `1.625 bpw` ternary packing scheme

In that case, it might be better to wait for this type instead of adding a temporary type to ggml that will become obsolete very quickly.

Reply:
Since the new datatype is still remain uncertainity, I'm thinking that I can simply drop the new datatype in this pr, and only preserve the bitnet model support, so that others can try new features or new datatypes (e.g. 1.625bpw) in llama.cpp to this model conveniently. It's this thought an acceptable one? @ggerganov @compilade @slaren

Reply:
> Since the new datatype is still remain uncertainity,

The `1.625 bpw` type is a work-in-progress in branch [`compilade/bitnet-ternary`](<https://github.com/ggerganov/llama.cpp/tree/compilade/bitnet-ternary>). It's called `Q1_3` for now, but the name will likely change, to follow the appropriate naming conventions. It works, and it's similar throughput to `Q4_0`. But it's slower than `Q2_2`. I think the performance ceiling is higher for the `2 bpw` quant, because of alignment, and because shifts and masks are cheaper on binary hardware than multiplications by powers of 3.

<details><summary>Performance of Q2_2 vs Q1_3 (click to expand)</summary>

---
(The tests below were done on an [Intel Core m3-8100Y](https://www.intel.com/content/www/us/en/products/sku/185282/intel-core-m38100y-processor-4m-cache-up-to-3-40-ghz/specifications.html), a low-power laptop CPU)

This is using commit <https://github.com/ggerganov/llama.cpp/commit/c6d9b115fd9899d072d809f2b2def51d5c63c3c9> from my work-in-progress branch with ternary quant types.

Q2_2 has been through some layout changes to make the performance higher and the AVX2 vec_dot simpler.

Not sure if I'll find more optimizations for Q1_3 though. I think it will always be slower than Q2_2. But I still hope to make the gap smaller.

```console
$ ./bin/test-quantize-perf --type q1_3 -i 10000000 --op vec_dot_q
q1_3
  vec_dot_q
    4096 values (0.02 MB)
      min cycles/32 vals   :      3.17
      avg cycles/32 vals   :      3.38
      float32 throughput   :     50.94 GB/s
      quantized throughput :      2.59 GB/s

$ ./bin/test-quantize-perf --type q2_2 -i 10000000 --op vec_dot_q
q2_2
  vec_dot_q
    4096 values (0.02 MB)
      min cycles/32 vals   :      2.18
      avg cycles/32 vals   :      2.35
      float32 throughput   :     69.66 GB/s
      quantized throughput :      4.35 GB/s

$ ./bin/test-quantize-perf --type q4_0 -i 10000000 --op vec_dot_q
q4_0
  vec_dot_q
    4096 values (0.02 MB)
      min cycles/32 vals   :      2.93
      avg cycles/32 vals   :      3.17
      float32 throughput   :     53.75 GB/s
      quantized throughput :      7.56 GB/s

$ ./bin/test-quantize-perf --type q8_0 -i 10000000 --op vec_dot_q
q8_0
  vec_dot_q
    4096 values (0.02 MB)
      min cycles/32 vals   :      2.31
      avg cycles/32 vals   :      2.42
      float32 throughput   :     68.19 GB/s
      quantized throughput :     18.11 GB/s

$ ./bin/test-quantize-perf --type q4_K -i 10000000 --op vec_dot_q
q4_K
  vec_dot_q
    4096 values (0.02 MB)
      min cycles/32 vals   :      2.40
      avg cycles/32 vals   :      2.59
      float32 throughput   :     64.15 GB/s
      quantized throughput :      9.02 GB/s

```

---

</details>

Should there be a choice between performance and compactness?
(Ideally, `1.625 bpw` would be the fastest, but this might not be possible)

> I can simply drop the new datatype in this pr, and only preserve the bitnet model support

Since this is a model architecture which is pretty much defined by its ability to be quantized to very low bpw, I think it's relevant to make the low bpw quant type(s) available along with it, although it's true that BitNet support doesn't strictly need these low bpw types.

Reply:
Briliant work on the Q2_2 and Q1_3 👍  I'm agree that he performance ceiling is higher for the `2 bpw` quant, the alignment issue indeed make 3bit quantize more complicated.

`Since this is a model architecture which is pretty much defined by its ability to be quantized to very low bpw, I think it's relevant to make the low bpw quant type(s) available along with it, although it's true that BitNet support doesn't strictly need these low bpw types.`

Also needs to note that, seems for now Q1_3 and Q2_2 only suits for the BitNet, or suits for the (-1, 0, 1) pattern weight(especially Q1_3, Q2_2 may suits 4 value weight). I personally suggests we can pull the Bitnet with no new datatypes in the master branch, and new datatypes like Q1_3 and Q2_2 in other branch, is all up to the llama.cpp team.

Reply:
@Eddie-Wang1120 I agree that we can merge initial model support with the existing data types and from there explore the best way to quantize

Regarding the datatype - I think it's interesting to explore the proposed `1.625 bpw` strategy. The `Q2_2` approach seems quite trivial to implement, but not sure yet if it is worth the extra code and maintenance associated with it. If bitnets become more widespread, we can reconsider

Reply:
Ok, I'll change this pr into a version without new datatype, and the new datatypes can be used through commit ID which I will show in pr intro.